### ISSUE TITLE
Compiler complains about an outer NotNullWhen inside a local function

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#42199]
**Version Used**: master

**Steps to Reproduce**:

```cs
using System;
using System.Diagnostics.CodeAnalysis;
#nullable enable
class C
{
    public int P { get; }
    
    private static bool Test([NotNullWhen(true)] out C? outParam)
    {
        outParam = null;   
        return false;
        static bool localFunc(C c)
            => c.P == 0 || // warns (none expected)
               c.P == 1;
    }
}
```

([sharplab repro](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFCEAMABIaQHQAimEA5gHYD2smAYyhMAwjwAmcAIJcIAG1xRMUANxUAxFwCucuRGBy4dOLINwqJOiKoBvKnQcMAzHUxcYdAAp0bdDnBgVOgBfe0cwhwAHBEw0CHgGUgA2OmAeHjk6ABU4WAAKCMcHAG0AOR4YUp05AHUACxM8mAQtOABKAF06Hi0PEQB+bt7PCAQIMDbCu0oiop6YEbGwOgBeOm1dIIdCosIAdjpseSg4NUodx0YUtIy6OR4BCEIAVjyROgFJmdmflYA+D5MbwrNb0fD4C4/RwCIGrNakM5FUKUZF0IA))

**Expected Behavior**: no warning about `outParam` in `localFunc`

**Actual Behavior**: `warning CS8762: Parameter 'outParam' may not have a null value when exiting with 'true'.`
